rstudio/rstudio
# Streamline VS Code + Docker Dev Experience

### Intent

To prepare for the Platform team's upcoming launcher work, I've been experimenting with ways to streamline setting up a dev environment.

The intent is not to change normal dev practices, but to introduce a streamlined way for occasional RSP developers to quickly ramp up without a complex dev environment setup. This PR is a result of that work, and it includes:

* A new readme (`VSCODE.md`) for getting started with Docker + VS Code
* Modifies the `.devcontainer` Dockerfile to invoke the dependency install scripts individually. This has two benefits:
  * Allows you to simply rebuild your Docker container periodically when dependencies change.
  * Calling scripts individually results in better layering; a change in one dependency doesn't require re-installing all dependencies.
* Sets `RSTUDIO_TARGET=Server` by default so configuration can be run without changes in the devcontainer approach. This does not affect configuration in other environments.
* Makes the `Xmx` and `Xss` memory settings configurable (same defaults for normal development). For devcontainer/VSCode development, these settings are customized to 1000M/16M to make building in Docker more likely to succeed.
* Throttles parallel build jobs to `4` (results in `-j 4`) for better results when building on Docker.
* Sets the correct include paths so you can jump to Boost sources in VS Code.
* Adds support for a `GWT_BUILD_ALL` env var. When set to `NO`, this suppresses the GWT build in the `all` target. I've found that the GWT build often fails when running in parallel with C++ builds. This is automatically suppressed in the devcontainer, but unaffected for other dev approaches.
* Adds a couple of VS Code plugins for auto-install in the devcontainer.

### Approach

The above changes should not affect other development environments like developing in a VM, developing locally on Linux or Mac/Windows, CI builds, or building with the docker build script. It only affects development when cloning the repo in a container volume using Docker + VS Code.

With the changes in this PR, I've been able to open this repo in VS Code and build it successfully with no configuration changes or failures.

### Caveats

* Makes some assumptions (like `-j 4`) that may not be ideal for everyone developing in Docker. I tried to find reasonable defaults that should work in most cases.
* The Dockerfile changes for better dependency layering may require a bit of attention from time to time, in particular if the dependencies installed by `install-common` change.
* The GWT build must be run separately when in the VS Code devcontainer environment. Personally, I prefer this, and I expect it's fine, but I could be wrong.

### Questions

I'm unclear on the current Java requirements. I ran into trouble when installing both Java 8 and 11, but you can resolve it by configuring a `JAVA_HOME` for the devcontainer. VS Code is unhappy about Java 8, but likes Java 11. Since GWT seems to build fine with Java 11, I simply removed Java 8 to simply things. Is this ok?

### Test Notes

To test with this branch:
* Make sure your Git SSH Key is in your SSH agent.
* Install the `Remote - Containers` VS Code extension.
* Choose "Remote-Containers: Clone Repository in Named Container Volume" from the command palette.
* Enter `git@github.com:/rstudio/rstudio-pro/tree/jon/vscode`.

### Checklist
